scipio/src/io/buffered_file.rs

```diff
@@ -0,0 +1,278 @@
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use scipio::LocalExecutor;
+    /// use scipio::io::BufferedFile;
+    /// use std::os::unix::io::AsRawFd;
+    ///
+    /// let ex = LocalExecutor::make_default();
+    /// ex.run(async {
+    ///     let mut wfile = BufferedFile::create("myfile.txt").await.unwrap();
+    ///     let mut rfile = BufferedFile::open("myfile.txt").await.unwrap();
+    ///     // Different objects (OS file descriptors), so they will be different...
+    ///     assert_ne!(wfile.as_raw_fd(), rfile.as_raw_fd());
+    ///     // However they represent the same object.
+    ///     assert!(wfile.is_same(&rfile));
+    ///     wfile.close().await;
+    ///     rfile.close().await;
+    /// });
+    /// ```
+    pub fn is_same(&self, other: &BufferedFile) -> bool {
+        self.file.is_same(&other.file)
+    }
+
+    /// Similar to [`create`] in the standard library, but returns a BufferedFile
+    ///
+    /// [`create`]: https://doc.rust-lang.org/std/fs/struct.File.html#method.create
+    pub async fn create<P: AsRef<Path>>(path: P) -> io::Result<BufferedFile> {
+        let flags = libc::O_CLOEXEC | libc::O_CREAT | libc::O_TRUNC | libc::O_WRONLY;
+        Ok(BufferedFile {
+            file: enhanced_try!(
+                ScipioFile::open_at(-1 as _, path.as_ref(), flags, 0o644).await,
+                "Creating",
+                Some(path.as_ref()),
+                None
+            )?,
+        })
+    }
+
+    /// Similar to [`open`] in the standard library, but returns a BufferedFile
+    ///
+    /// [`open`]: https://doc.rust-lang.org/std/fs/struct.File.html#method.open
+    pub async fn open<P: AsRef<Path>>(path: P) -> io::Result<BufferedFile> {
+        let flags = libc::O_CLOEXEC | libc::O_RDONLY;
+        Ok(BufferedFile {
+            file: enhanced_try!(
+                ScipioFile::open_at(-1 as _, path.as_ref(), flags, 0o644).await,
+                "Reading",
+                Some(path.as_ref()),
+                None
+            )?,
+        })
+    }
+
+    /// Write the data in the buffer `buf` to this BufferedFile at the specified position
+    pub async fn write_at(&self, buf: &[u8], pos: u64) -> io::Result<usize> {
+        let source = Reactor::get().write_buffered(self.as_raw_fd(), buf, pos);
+        enhanced_try!(source.collect_rw().await, "Writing", self.file)
+    }
+
+    /// Reads data at the specified position into the user-provided buffer `buf`.
+    ///
+    /// Note that this differs from [`DmaFile`]'s read APIs: that reflects the
+    /// fact that buffered reads need no specific alignment and io_uring will not
+    /// be able to use its own pre-allocated buffers for it anyway.
+    ///
+    /// [`DmaFile`]: struct.DmaFile.html
+    pub async fn read_at(&self, buf: &mut [u8], pos: u64) -> io::Result<usize> {
+        let source = Reactor::get().read_buffered(self.as_raw_fd(), buf, pos);
+        enhanced_try!(source.collect_rw().await, "Reading", self.file)
+    }
```

Comment:
Hm....

I think this API is fundamentally unsound. If the user cancells this `read_at` operation (by dropping the future), wouldn't consider `buf` mutably borrowed anymore, but the kernel would happily write into it, until we *await* for cancellation. And we can't await for cancellation in Drop because drop is sync, and because that would  be unsound anyway due to mem::forget or rc cycles. 

See https://without.boats/blog/io-uring/ and https://without.boats/blog/ringbahn/ for some discussion of this problem.

I think the fix would be to incur a copy -- we should use our owned DMA buffer internally, and copy the data to a user-provided buf when we are done. 

(note -- this is cursory glance at the code, maybe I am mistaken here)

Reply:
I am seriously unhappy about the amount of copies that are forced in the read path by the APIs. There are some in the `AsyncReadBuf` too... 
But you are probably right. In which case it would probably better to have an API more similar to the DmaFile, where we allocate our own buffers. They would just be simple buffers instead of the DmaBuffer with their alignment guarantees.

Reply:
>But you are probably right. In which case it would probably better to have an API more similar to the DmaFile, where we allocate our own buffers. They would just be simple buffers instead of the DmaBuffer with their alignment guarantees.

Yeah, agree. I think the simplest possible fix here is to just use the following API 

```rust
pub async fn read_at(&self, pos: u64, len: usize) -> io::Result<Vec<u8>> {
```

Long term, we should replace `Vec<u8>` with some dedicated type, which represents a buffer from a buffer pool owned by scipio.

Reply:
This is a bit worse, I think.
I can't convince myself write wouldn't have the same problem.

But that also means that the current interface for `write_dma` is unsound.
We're taking a `DmaBuffer` reference as a parameter there. The best solution is probably to just take ownership of the buffer and store it in the `Source`. We keep the `Source` alive throughout the whole operation, so that's safe.

And for our `write_at` here in the `BufferedFile`, same thing: pass the ownership of a vector, not a reference...

Ok, this will be a bit of a big structural change. In the end both APIs will be very similar, which is good news.

I still want to keep them separate as APIs because whether or not you are using Direct I/O is a big decision, shouldn't be swept under the hood. But common implementations will help.